extensions/amp-bind/0.1/amp-state.js

```diff
@@ -207,15 +211,39 @@ export class AmpState extends AMP.BaseElement {
   /**
    * @param {boolean} isInit
    * @param {boolean=} opt_refresh
-   * @return {!Promise<undefined>}
+   * @return {!Promise}
    * @private
    */
   fetchAndUpdate_(isInit, opt_refresh) {
+    // On init, we reuse the deferred created in the constructor.
+    if (!isInit) {
+      this.loadingDeferred_ = new Deferred();
+    }
+
+    // Store the deferred locally, in case a new fetch overwrites
+    // it before resolution.
+    const loadingDeferred = this.loadingDeferred_;
+
     // Don't fetch in prerender mode.
     return this.getAmpDoc()
       .whenFirstVisible()
       .then(() => this.prepareAndSendFetch_(isInit, opt_refresh))
-      .then(json => this.updateState_(json, isInit));
+      .then(json => this.updateState_(json, isInit))
+      .then(() => loadingDeferred.resolve());
+  }
+
+  /**
+   * For an "amp-state" with a src attribute, this returns a promise that
+   * resolves after the fetch and update completes. For non-src "amp-state"s,
+   * return a resolved promise.
+   *
+   * @return {!Promise}
+   */
+  getFetchingPromise() {
+    if (!this.element.hasAttribute('src')) {
+      return Promise.resolve();
+    }
+    return this.loadingDeferred_.promise;
```

Comment:
Optional: I wonder if resolving the Promise with the data (or `undefined` in no `src` case) would be a bit simpler.

Reply:
I think it would be a tiny bit simpler, but skipping `getState` would would make me slightly uncomfortable.
I'm also not sure it would technically be correct since we deep merge into a key, and technically there could be pre-existing data within a key which the promise wouldn't capture.